--- ../../Xperia-CM/cm11.0/twrp/android/system/bootable/recovery/minuitwrp/graphics.c	2015-07-20 09:14:39.671928699 +0200
+++ recovery/graphics/twrpgraphics.c	2015-07-20 11:30:23.736025006 +0200
@@ -54,6 +54,7 @@
 
 #define NUM_BUFFERS 2
 #define MAX_DISPLAY_DIM  2048
+#define FB_BOUND 0X1000
 
 // #define PRINT_SCREENINFO 1 // Enables printing of screen info to log
 
@@ -295,9 +296,14 @@
 
     fb++;
 
+    unsigned int fb_size = vi.yres * fi.line_length;
+    if (fb_size % FB_BOUND != 0) {
+        fb_size += FB_BOUND - fb_size % FB_BOUND;
+    }
+
 #ifndef TW_DISABLE_DOUBLE_BUFFERING
     /* check if we can use double buffering */
-    if (vi.yres * fi.line_length * 2 > fi.smem_len)
+    if (fb_size * 2 > fi.smem_len)
 #else
     printf("TW_DISABLE_DOUBLE_BUFFERING := true\n");
 #endif
@@ -313,7 +319,7 @@
     fb->data = (GGLubyte*) (((unsigned long) bits) + vi.yres * fi.line_length);
 #else
     fb->stride = vi.xres_virtual;
-    fb->data = (GGLubyte*) (((unsigned long) bits) + vi.yres * fb->stride * PIXEL_SIZE);
+    fb->data = (void*) (((unsigned) bits) + fb_size);
 #endif
     fb->format = PIXEL_FORMAT;
     if (!has_overlay) {
